vpr/src/place/noc_place_utils.h

```diff
@@ -0,0 +1,375 @@
+constexpr double MAX_INV_NOC_AGGREGATE_BANDWIDTH_COST = 1.;
+// we expect the latency costs to be in the picosecond range, and we don't expect it to go lower than that. So if the latency costs go below the picosecond range we trim the normalization value to below 1 picosecond
+// This should be updated if the delays become lower
+constexpr double MAX_INV_NOC_LATENCY_COST = 1.e12;
+
+// we don't expect the noc_latency cost to ever go below 1 pico second.
+// So this value represents the lowest possible latency cost.
+constexpr double MIN_EXPECTED_NOC_LATENCY_COST = 1.e-12;
+
+/* Stores statistical data about how the NoC blocks were moved during placement
+ */
+struct NocPlaceStats {
+    int number_of_noc_router_moves;
+    std::vector<int> number_of_noc_router_moves_per_move_type;
+};
+
+/* Defines how the links found in a traffic flow are updated in terms
+ * of their bandwidth usage.
+ */
+enum class link_usage_update_state {
+    /* State 1: The link usages have to be incremented as the traffic
+     * flow route route has been updated 
+     */
+    increment,
+    /* State 2: The link usages have to be decremented as the traffic flow
+     * route is being removed
+     */
+    decrement
+};
+
+/**
+ * @brief Routes all the traffic flows within the NoC and updates the link usage
+ * for all links. This should be called after initial placement, where all the 
+ * logical NoC router blocks have been placed for the first time and no traffic
+ * flows have been routed yet. This function should also only be used once as
+ * its intended use is to initialize the routes for all the traffic flows.
+ * 
+ */
+void initial_noc_placement(void);
+
+/**
+ * @brief Goes through all the cluster blocks that were moved
+ * in a single swap iteration during placement and checks to see
+ * if any moved blocks were NoC routers. 
+ * 
+ * For each moved block that is a NoC router, all the traffic flows
+ * that the router is a part of are re-routed. The noc placement
+ * costs (latency and aggregate bandwdith) are also updated to
+ * reflect the re-routed traffic flows.
+ * 
+ * If none of the moved blocks are NoC routers, then this function
+ * does nothing.
+ * 
+ * This function should be used if the user enabled NoC optimization
+ * during placement and only if the move was accepted by the placer.
+ * 
+ * @param blocks_affected Contains all the blocks that were moved in
+ * the current placement iteration. THis includes the cluster ids of
+ * the moved blocks, their previous locations and their new locations
+ * after being moved.
+ */
+int find_affected_noc_routers_and_update_noc_costs(const t_pl_blocks_to_be_moved& blocks_affected, double& noc_aggregate_bandwidth_delta_c, double& noc_latency_delta_c, const t_noc_opts& noc_opts);
+
+void commit_noc_costs(int number_of_affected_traffic_flows);
```

Comment:
Maybe add a single line comment for this function.

Reply:
Updated comment